### PR TITLE
Optimise content type lookups in connectors

### PR DESCRIPTION
**Pull request by AndyButland**
We've a couple of connectors in here - for the block list and nested content - that need to look up content types.  This is done multiple times in the context of a back-office save or a content transfer to deploy operation.

To optimise this I've added some caching, so it'll only make the request once per operation.

- I'm using the request cache if available (as it is when saving in the back-office), as this is really the behaviour we want, to cache for the context of the request and not beyond that.
- For a content transfer though, this happens in a background thread and so the request cache isn't available.  So instead here I'm using the runtime cache.
    - There's [another PR for Deploy](https://github.com/umbraco/Umbraco-Deploy/pull/657) that clears this between operations, so again we don't cache the data for longer than we want to.

### CODE REVIEW

File: src/Umbraco.Deploy.Contrib.Connectors/ValueConnectors/BlockEditorValueConnector.cs [hunk 1]
```diff
@@ -129,7 +136,40 @@ public virtual string ToArtifact(object value, PropertyType propertyType, IColle
...
+
+            // If we've requested this set of content types within the request, get from cache rather than going to the database again.
+            // This is useful as this method can be called multiple times in a save/publish operation.
+            var cacheKey = string.Format(DeployContribConstants.CacheKeys.DeployConnectorCacheContentTypesFormat, string.Join(",", allBlockContentTypeKeys));
+
+            // If we're on a background thread (e.g. for a content transfer), the request cache won't exist and we would fall through to the database retrieval.
+            // So instead we'll use the runtime cache (which will be cleared by Deploy between each background operation).
+            Func<Dictionary<string, IContentType>> getContentTypes = () =>
+                allBlockContentTypeKeys
+                    .ToDictionary(x => x.ToString(), x => _contentTypeService.Get(x));
```
Comment (ronaldbarendse) [inline src/Umbraco.Deploy.Contrib.Connectors/ValueConnectors/BlockEditorValueConnector.cs:159]: This still fetches the content types one by one from the database, which can be avoided by using `GetAll(allBlockContentTypeKeys)` instead.

**Review by ronaldbarendse — CHANGES_REQUESTED** [review]
Looks like we can optimize the current code even more by using `GetAll()` instead of multiple `Get()`s. But looking at the `ContentTypeService`/`ContentTypeRepository`, these calls should already be cached using the `FullDataSetRepositoryCachePolicy`: https://github.com/umbraco/Umbraco-CMS/blob/c0c9c50e2110a88afab800abbf0e4a6b6a08c62d/src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentTypeRepository.cs#L33

Whether this cache gets used depends on the `RepositoryCacheMode` of the scope though. If that's set to `None`, each `Get()` would do a `PerformGetAll()` that fetches everything from the database and filter them in-memory...

In short: just changing `Get()` to `GetAll()` will improve performance massively if the repository cache isn't used. Otherwise we should make sure the repository cache is used instead of adding our own caching layer.

**Comment by AndyButland** [comment]
`GetAll` only has an overload taking an array of integer Ids unfortunately, which we don't have here - which is why I assume it was coded to retrieve one at a time in the first place.

We have `RepositoryCacheMode.Scoped` set on the scope Deploy creates, and having checked with SQL Profiler I'm actually not seeing multiple SQL requests going off here.  So I guess as you suggest this isn't going to help too much...
